### DISCUSSION TITLE
Nested each loops

### DISCUSSION THREAD

**Discussion by MarioIonchev** [cypress-io/cypress#29972] [Questions and Help]
Hi, guys. I`m trying to test a product with some attributes that are somehow linked if they are available. I have 4 nested loops and there maybe more in the future. What i`m trying to do is check if the first one`s attribute is available -> click() and continue with the same logic in the other nested loops. The last one checks for an available class and if (true) should -> click() and brake the loop. Due to asynchronous nature of Cypress, i`m not able to do it. I tried the cy.wrap() approach, but i get the error ( Element is detached from the DOM). To avoid this i`m using JQuery, Cypress.$ to get the element in each loop and click(). But when i log the index of each loop i see that they`re not working as nested loops, but are executed 1 by 1. Has any encounter such problem, or am i just doing something wrong here? Every workaround and solution would be appreciated. This is my current version of Cypress -> "cypress": "^11.2.0". Here`s my current code.

selectOptions() {
        let found = false;

        cy.get(newQualityArray).each(($quality, i) => {
            if (found) return false;
            Cypress.$($quality).trigger('click');

            cy.get(newWeightArray).each(($weight, i) => {
                if (found) return false;
                Cypress.$($weight).trigger('click');

                cy.get(newMetaltArray).each(($metal, i) => {
                    if (found) return false;
                    Cypress.$($metal).trigger('click');

                    cy.get(newSizeArray).each(($size, i) => {
                        if ($size.hasClass('selectable')) {
                            if (!found) {
                                Cypress.$($size).trigger('click');
                                found = true;
                                return false;
                            }
                        }
                    });
                });
            });
        });
    }

**Comment by portyu9** [comment]
Hi @MarioIonchev 

The main problem is the combination of **Cypress's queued commands** with synchronous nested loops and a synchronous `found` variable.

Code such as:

```js
cy.get(...).each(() => {
  cy.get(...).each(() => {
    // ...
    found = true
  })

  if (found) {
    // ...
  }
})
```

does not execute like normal nested synchronous JavaScript. The outer `.each()` callback queues Cypress commands, and `found` is only changed later when those queued commands execute.

There is a second issue here:

```js
Cypress.$($el).trigger('click')
```

bypasses Cypress's normal `.click()` behavior, actionability checks, retries, and command sequencing. If clicking an option causes the application to re-render the available options, previously captured DOM elements can also become detached/stale.

For a search such as "try each option until a selectable combination is found", I would make the traversal explicitly sequential and re-query the DOM on each iteration.

For example:

```js
function findFirst(selector, tryItem, index = 0) {
  return cy.get(selector).then(($items) => {
    if (index >= $items.length) {
      return false
    }

    return cy.get(selector)
      .eq(index)
      .then(($item) => {
        return tryItem($item).then((found) => {
          if (found) {
            return true
          }

          return findFirst(selector, tryItem, index + 1)
        })
      })
  })
}
```

Then the nested search can be expressed as returned Cypress chains:

```js
findFirst(newQualityArray, ($quality) => {
  return cy.wrap($quality).click().then(() => {
    return findFirst(newWeightArray, ($weight) => {
      return cy.wrap($weight).click().then(() => {
        return findFirst(newMetaltArray, ($metal) => {
          return cy.wrap($metal).click().then(() => {
            return cy.get(newSizeArray).then(($sizes) => {
              const $selectable = $sizes.filter('.selectable').first()

              if (!$selectable.length) {
                return false
              }

              return cy.wrap($selectable)
                .click()
                .then(() => true)
            })
          })
        })
      })
    })
  })
})
```

The important parts are:

- return each Cypress chain so the next level waits;
- do not use a normal boolean to coordinate work that runs later in Cypress's command queue;
- use `cy.wrap(...).click()` rather than a raw jQuery `trigger('click')`;
- re-query collections after an action if the application's DOM can re-render.

If changing a parent selection does not automatically reset the child selections in your application, you will also need to explicitly reset those before trying the next parent combination.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
